include/openthread/thread_ftd.h

```diff
@@ -580,47 +580,65 @@ OTAPI int8_t OTCALL otThreadGetParentPriority(otInstance *aInstance);
 OTAPI otError OTCALL otThreadSetParentPriority(otInstance *aInstance, int8_t aParentPriority);
 
 /**
- * This enumeration defines the constants used in `otThreadChildTableCallback` to indicate whether a child is added or
- * removed.
+ * This enumeration defines the constants used in `otNeighborTableCallback` to indicate whether a child or router
+ * neighbor is being added or removed.
  *
  */
-typedef enum otThreadChildTableEvent
+typedef enum
 {
-    OT_THREAD_CHILD_TABLE_EVENT_CHILD_ADDED,   ///< A child is being added.
-    OT_THREAD_CHILD_TABLE_EVENT_CHILD_REMOVED, ///< A child is being removed.
-} otThreadChildTableEvent;
+    OT_NEIGHBOR_TABLE_EVENT_CHILD_ADDED,    ///< A child is being added.
+    OT_NEIGHBOR_TABLE_EVENT_CHILD_REMOVED,  ///< A child is being removed.
+    OT_NEIGHBOR_TABLE_EVENT_ROUTER_ADDED,   ///< A router is being added.
+    OT_NEIGHBOR_TABLE_EVENT_ROUTER_REMOVED, ///< A router is being removed.
+} otNeighborTableEvent;
 
 /**
- * This function pointer is called to notify that a child is being added to or removed from child table.
+ * This type represent a neighbor table entry info (child or router) and is used as a parameter in the neighbor table
+ * callback `otNeighborTableCallback`.
  *
- * @param[in]  aEvent      A event flag indicating whether a child is being added or removed.
- * @param[in]  aChildInfo  A pointer to child information structure.
+ */
+typedef struct
+{
+    otInstance *mInstance; // The OpenThread instance.
+    union
+    {
+        otChildInfo    mChild;  ///< The child neighbor info.
+        otNeighborInfo mRouter; ///< The router neighbor info.
+    } mInfo;
+} otNeighborTableEntryInfo;
+
+/**
+ * This function pointer is called to notify that a child or router neighbor is being added to or removed from neighbor
+ * table.
+ *
+ * @param[in]  aEvent      A event flag.
+ * @param[in]  aEntryInfo  A pointer to table entry info.
  *
  */
-typedef void (*otThreadChildTableCallback)(otThreadChildTableEvent aEvent, const otChildInfo *aChildInfo);
+typedef void (*otNeighborTableCallback)(otNeighborTableEvent aEvent, const otNeighborTableEntryInfo *aEntryInfo);
```

Comment:
Pass a `void*` for higher-layer context?

Reply:
Talked offline with Jonathan. Plan is to leave this to caller to save any state/context it may want, mainly since we are providing `mInstance` in the entry and the caller can use this to determine which instance is invoking the callback.